OPAL/tac/src/main/scala/org/opalj/tac/cg/android/AndroidICCAnalysis.scala

```diff
@@ -0,0 +1,1467 @@
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        }
+        //adding explicit intent edges
+        explicitIntents.foreach { i ⇒
+            val caller = i.caller
+            i.calledMethod.foreach { m ⇒
+                edges += ((caller, i.pc) -> m)
+            }
+        }
+    }
+
+    /**
+     * Test whether the data field of an intent matches the data field of any of the filters.
+     * If there is a match, an edge is created.
+     * @param intent Intent that provides the data that is tested.
+     * @param filters ListBuffer of IntentFilters that are matched against the intent.
+     */
+    def testIntentDataAgainstFilters(
+        intent:  ImplicitIntent,
+        filters: ListBuffer[IntentFilter]
+    ): Unit = {
+        for (filter ← filters) {
+            //The data of the intent is split up in its parts to be tested against the filters.
+            val types = filter.dataTypes
+            val schemes = filter.dataSchemes
+            val data = intent.iData
+            if (types.isEmpty && schemes.isEmpty && intent.iType.isEmpty && data.isEmpty) {
+                matchFound(intent, filter)
+            } else if (data.nonEmpty) {
+                val ssi = data.indexOf(":")
+                val scheme = data.substring(0, ssi)
+                if (schemes.contains(scheme)) {
+                    val authorities = filter.dataAuthorities
+                    val authorityEnd = findAuthority(data)
+                    if (authorityEnd > ssi + 3 && authorities.contains(data.substring(ssi + 3, authorityEnd))) {
+                        val paths = filter.dataPaths
+                        if (paths.isEmpty) {
+                            matchFound(intent, filter)
+                        } else if (ssi < authorityEnd && paths.exists(p ⇒ p.startsWith(findPath(data, authorityEnd)))) {
+                            val types = filter.dataTypes
+                            val iType = intent.iType
+                            if (types.isEmpty && iType.isEmpty || types.contains(iType)) {
+                                matchFound(intent, filter)
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    def matchFound(
+        intent: ImplicitIntent,
+        filter: IntentFilter
+    ): Unit = {
+        val caller = intent.caller
+        findCalledMethods(filter.receiver, intent.calledMethods, intent.componentTypes).foreach { m ⇒
+            edges += ((caller, intent.pc) -> m)
+        }
+    }
+
+    def findAuthority(
+        uri: String
+    ): Integer = {
+        val ssi = uri.indexOf(":")
+        if (uri.substring(ssi).startsWith("//")) {
+            for (index ← ssi + 3 to uri.length) {
+                uri.charAt(index) match {
+                    case '/' | '\\' | '?' | '#' ⇒ return index;
```

Comment:
Wouldn't this result in a match not exhaustive error if the character is anything else?

Reply:
Also, isn't there some kind of `indexOf` method that could be used here instead of building this your own?